reference/solver/batch_richardson_kernels.cpp

```diff
@@ -0,0 +1,257 @@
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/solver/batch_richardson_kernels.hpp"
+
+
+#include "core/matrix/batch_struct.hpp"
+#include "reference/base/config.hpp"
+#include "reference/log/batch_logger.hpp"
+// include device kernels for every matrix and preconditioner type
+#include "reference/matrix/batch_csr_kernels.hpp"
+#include "reference/matrix/batch_dense_kernels.hpp"
+#include "reference/matrix/batch_struct.hpp"
+#include "reference/preconditioner/batch_jacobi.hpp"
+#include "reference/stop/batch_criteria.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace reference {
+
+
+#define GKO_REF_BATCH_USE_DYNAMIC_SHARED_MEM 1
+
+/**
+ * @brief The batch Richardson solver namespace.
+ *
+ * @ingroup batch_rich
+ */
+namespace batch_rich {
+
+
+template <typename T>
+using BatchRichardsonOptions =
+    gko::kernels::batch_rich::BatchRichardsonOptions<T>;
+
+template <typename StopType, typename PrecType, typename LogType,
+          typename BatchMatrixType, typename ValueType>
+static void apply_impl(
+    std::shared_ptr<const ReferenceExecutor>,
+    const BatchRichardsonOptions<remove_complex<ValueType>> &opts,
+    LogType logger, PrecType prec, const BatchMatrixType &a,
+    const gko::batch_dense::UniformBatch<const ValueType> &left_scale,
+    const gko::batch_dense::UniformBatch<const ValueType> &right_scale,
+    const gko::batch_dense::UniformBatch<ValueType> &b,
+    const gko::batch_dense::UniformBatch<ValueType> &x)
+{
+    using real_type = typename gko::remove_complex<ValueType>;
+    const size_type nbatch = a.num_batch;
+    const auto nrows = a.num_rows;
+    const auto nrhs = b.num_rhs;
+    constexpr int max_nrhs = batch_config<ValueType>::max_num_rhs;
+    constexpr int max_nrows = batch_config<ValueType>::max_num_rows;
+
+    GKO_ASSERT((nrhs == x.num_rhs));
+    GKO_ASSERT((max_nrows * max_nrhs >= nrows * nrhs));
+
+    for (size_type ibatch = 0; ibatch < nbatch; ibatch++) {
+#if GKO_REF_BATCH_USE_DYNAMIC_SHARED_MEM
+        const int shared_size_bytes =
+            gko::kernels::batch_rich::local_memory_requirement<ValueType>(
+                nrows, nrhs) +
+            PrecType::dynamic_work_size(nrows, a.num_nnz) * sizeof(ValueType);
+        const auto shared_space = malloc(shared_size_bytes);
+        const auto residual = reinterpret_cast<ValueType *>(shared_space);
+        ValueType *const delta_x = residual + nrows * nrhs;
+        ValueType *const prec_work = delta_x + nrows * nrhs;
+        const auto norms = reinterpret_cast<real_type *>(
+            prec_work + PrecType::dynamic_work_size(nrows, a.num_nnz));
+        real_type *const init_rel_res_norm = norms + nrhs;
+#else
+        ValueType residual[max_nrows * max_nrhs];
+        ValueType delta_x[max_nrows * max_nrhs];
+        ValueType prec_work[PrecType::work_size];
+        real_type norms[max_nrhs];
+        real_type init_rel_res_norm[max_nrhs];
+#endif
```

Comment:
I think if this is not needed for reference, then we should just remove it and simplify the code here.

Reply:
I was thinking of keeping the reference code kind of consistent with cuda code with regard to dynamic memory usage, so maybe we could remove this once the question is settled for Cuda? (But you're right that apart from a vague consistency argument, there's no real reason for this.)

Reply:
In general, I think there is no need to be consistent with CUDA. Actually, it is better to have code that is as simple as possible in reference. If there is some bug in CUDA and reference uses code similar to that, then this makes things complicated to test and debug.

Reply:
Then I would probably just keep the dynamic allocation version here. Then we could test any matrix size.